File: frontend/src/services/logs.js
import { api } from './api'

export const ENTIDADES_LOG = [
  { value: 'maquina', label: 'Máquina' },
  { value: 'centro_despesa', label: 'Centro de Despesa' },
  { value: 'fornecedor', label: 'Fornecedor' },
  { value: 'manutencao', label: 'Manutenção' },
  { value: 'preventiva', label: 'Preventiva' },
  { value: 'abastecimento', label: 'Abastecimento' },
  { value: 'nota_fiscal', label: 'Nota Fiscal' },
  { value: 'checklist_execucao', label: 'Checklist' },
  { value: 'usuario', label: 'Usuário' },
]

export const ACOES_LOG = [
  { value: 'criar', label: 'Criação' },
  { value: 'atualizar', label: 'Atualização' },
  { value: 'excluir', label: 'Exclusão' },
  { value: 'excluir_permanente', label: 'Exclusão permanente' },
  { value: 'ativar', label: 'Ativação' },
  { value: 'desativar', label: 'Desativação' },
]

export async function listarLogs(filtros = {}) {
  const params = {}
  if (filtros.entidade) params.entidade = filtros.entidade
  if (filtros.acao) params.acao = filtros.acao
  if (filtros.usuario_id) params.usuario_id = filtros.usuario_id
  if (filtros.data_inicio) params.data_inicio = filtros.data_inicio
  if (filtros.data_fim) params.data_fim = filtros.data_fim
  if (filtros.busca) params.busca = filtros.busca
  const { data } = await api.get('/logs', { params })
  return data
}

export async function listarUsuariosComLog() {
  const { data } = await api.get('/logs/usuarios-ativos')
  return data
}

export function labelEntidade(valor) {
  return ENTIDADES_LOG.find((e) => e.value === valor)?.label || valor
}

export function labelAcao(valor) {
  return ACOES_LOG.find((a) => a.value === valor)?.label || valor
}

File: backend/app/routers/logs.py
from datetime import date, datetime, timedelta
from typing import Optional

from fastapi import APIRouter, Depends, Query

from app.auth_deps import exigir_admin, UsuarioLogado
from app.database import get_supabase

router = APIRouter(prefix="/logs", tags=["Central de Logs"])

ENTIDADES = [
    "maquina", "centro_despesa", "fornecedor", "manutencao", "preventiva",
    "abastecimento", "nota_fiscal", "checklist_execucao", "usuario",
]
ACOES = ["criar", "atualizar", "excluir", "excluir_permanente", "ativar", "desativar"]


@router.get("")
def listar_logs(
    entidade: Optional[str] = None,
    acao: Optional[str] = None,
    usuario_id: Optional[int] = None,
    data_inicio: Optional[date] = None,
    data_fim: Optional[date] = None,
    busca: Optional[str] = Query(None, description="Busca por texto na descrição"),
    limite: int = Query(200, ge=1, le=1000),
    _: UsuarioLogado = Depends(exigir_admin),
):
    """Lista o histórico de ações do sistema -- somente administradores."""
    sb = get_supabase()
    query = sb.table("logs_auditoria").select("*")

    if entidade:
        query = query.eq("entidade", entidade)
    if acao:
        query = query.eq("acao", acao)
    if usuario_id:
        query = query.eq("usuario_id", usuario_id)
    if data_inicio:
        query = query.gte("created_at", data_inicio.isoformat())
    if data_fim:
        # inclui o dia inteiro de data_fim
        query = query.lt("created_at", (data_fim + timedelta(days=1)).isoformat())
    if busca:
        query = query.ilike("descricao", f"%{busca}%")

    resp = query.order("created_at", desc=True).limit(limite).execute()
    return resp.data


@router.get("/usuarios-ativos")
def listar_usuarios_com_log(_: UsuarioLogado = Depends(exigir_admin)):
    """Lista os usuários distintos que aparecem no log, para popular o filtro por usuário."""
    sb = get_supabase()
    resp = sb.table("logs_auditoria").select("usuario_id, usuario_nome").execute()
    vistos = {}
    for linha in resp.data:
        if linha["usuario_id"] is not None:
            vistos[linha["usuario_id"]] = linha["usuario_nome"]
    return [{"id": uid, "nome": nome} for uid, nome in sorted(vistos.items(), key=lambda x: x[1])]
